detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/DoubleMutabilityForCollection.kt

```diff
@@ -0,0 +1,79 @@
+package io.gitlab.arturbosch.detekt.rules.bugs
+
+import io.gitlab.arturbosch.detekt.api.CodeSmell
+import io.gitlab.arturbosch.detekt.api.Config
+import io.gitlab.arturbosch.detekt.api.Debt
+import io.gitlab.arturbosch.detekt.api.Entity
+import io.gitlab.arturbosch.detekt.api.Issue
+import io.gitlab.arturbosch.detekt.api.Rule
+import io.gitlab.arturbosch.detekt.api.Severity
+import org.jetbrains.kotlin.descriptors.VariableDescriptor
+import org.jetbrains.kotlin.js.descriptorUtils.getJetTypeFqName
+import org.jetbrains.kotlin.psi.KtProperty
+import org.jetbrains.kotlin.resolve.BindingContext
+
+/**
+ * Using `var` when declaring a mutable collection leads to double mutability. Consider instead
+ * declaring your variable with `val` or switching your declaration to use an immutable type.
+ *
+ * <noncompliant>
+ * var myList = mutableListOf(1,2,3)
+ * var mySet = mutableSetOf(1,2,3)
+ * var myMap = mutableMapOf("answer" to 42)
+ * </noncompliant>
+ *
+ * <compliant>
+ * // Use val
+ * val myList = mutableListOf(1,2,3)
+ * val mySet = mutableSetOf(1,2,3)
+ * val myMap = mutableMapOf("answer" to 42)
+ *
+ * // Use immutable types
+ * var myList = listOf(1,2,3)
+ * var mySet = setOf(1,2,3)
+ * var myMap = mapOf("answer" to 42)
+ * </compliant>
+ *
+ * @requiresTypeResolution
+ */
+class DoubleMutabilityForCollection(config: Config = Config.empty) : Rule(config) {
+
+    override val issue: Issue = Issue(
+        "DoubleMutabilityInCollection",
+        Severity.CodeSmell,
+        "Using var with mutable collections leads to double mutability. " +
+                "Consider using val or immutable collection types.",
+        Debt.FIVE_MINS
+    )
+
+    override fun visitProperty(property: KtProperty) {
+        super.visitProperty(property)
+
+        if (bindingContext == BindingContext.EMPTY) return
+
+        val type =
+            (bindingContext[BindingContext.DECLARATION_TO_DESCRIPTOR, property] as? VariableDescriptor)?.type ?: return
+        val standardType = type.getJetTypeFqName(false)
```

Comment:
```suggestion
       val standardType = type.constructor.declarationDescriptor?.fqNameOrNull()?.asString()
```
`getJetTypeFqName` function is in the `js` package, so I think you shouldn't use it.

Reply:
🤔 should we add `org.jetbrains.kotlin.js.*` as a forbidden import to avoid this? It's not the first time that I see this kind of error.

Reply:
Sadly other rules are also using this util: https://github.com/detekt/detekt/search?q=getJetTypeFqName

I wasn't able to find an equivalent in the common compiler package.
Given that this function is using nothing js specific, I suggest we copy it over to the `-psi-utils`:

https://github.com/JetBrains/kotlin/blob/29b23e79f32791e456a5b4a453277f0f0b3e984d/js/js.frontend/src/org/jetbrains/kotlin/js/descriptorUtils/descriptorUtils.kt#L22-L41

Reply:
This is a good catch @t-kameyama and I have already raised the same concern in other PRs. There is no replacement for this function in the Kotlin source code.

I don't think we want to block this PR because it is using `getJetTypeFqName()`, I can help create a new issue to address this.

Reply:
https://github.com/detekt/detekt/issues/3605

Reply:
I would opt for merging this rule and following up on #3605 once solved.

Reply:
If this is merged I'll address it in #3613

Reply:
We need to update this PR since #3613 is merged.

Reply:
Never mind, I can submit another PR addressing this.